### ISSUE TITLE
New Tracking Patterns

### ISSUE THREAD

**Issue by damassi** [artsy/reaction#1386]
After being tasked with implementing tracking on a few areas of the app I got a chance to dive into our new tracking patterns established around the time the new /artist app was built and it got me thinking about the possibility that we're engineering a solution to something that might not be a problem and that there might be a complexity cost at scale. 

New pattern: 
https://github.com/artsy/reaction/blob/master/src/Artsy/Analytics/Schema/Values.ts

How its used: 

```js
@track({ 
  context_module: Schema.Context.MinimalCtaBanner 
})
export class MinimalCtaBanner extends React.Component {
  @track({
    action_type: Schema.ActionType.AuthImpression,
    intent: Schema.ActionName.ViewEditorial,
    trigger: "click",
  })
  componentDidMount() {
    // no op
  }
}
```

Old pattern: 
https://github.com/artsy/force/tree/master/src/desktop/analytics 
https://github.com/artsy/force/blob/master/src/desktop/analytics/artwork.js (example context)

How its used:

The Analytics team passes us a snippet: 

```js
analytics.track('Click', {
  flow: 'Artwork about the work',
  context_module: 'About the Work (Partner)',
  type: 'Button',
  label: 'Read more',
})
```

Which we would then paste in via `react-tracking`: 

```js
@track({
  flow: 'Artwork about the work',
  context_module: 'About the Work (Partner)',
  type: 'Button',
  label: 'Read more',
  trigger: "Click",
})
```

In the new system, we take the values that Analytics passes us and we encode them into enums under specific buckets grouped by type, and then we import those types into our code and pass into `@track()`. 

What are the benefits of type safety in regards to our tracking code?

Things like `ActionType` — repetitive bits that we see duplicated all throughout old analytics code — would benefit. That way we don’t have things like “Click” or “Clickk”, etc. Those properties can be encoded in an enum or variable and make sense. 

<img width="376" alt="screen shot 2018-10-09 at 5 03 06 pm" src="https://user-images.githubusercontent.com/236943/46707281-20ef2a80-cbee-11e8-8179-41981ea43280.png">

But when looking at things like `ActionName`, these are unique values:

<img width="354" alt="screen shot 2018-10-09 at 4 38 06 pm" src="https://user-images.githubusercontent.com/236943/46707297-35cbbe00-cbee-11e8-8377-c84a1f61e087.png">

What is the value in creating duplicate definitions of something that is unique? `dismiss` now has a slightly different mirror in `Dismiss`; `passwordNextButton` now has a mirror in `PasswordNextButton`, which is just more work to maintain over time, and even if they were matched 1:1, that’s still two identical values (`passwordNextButton = "passwordNextButton"`).

`Context` is another example:

<img width="365" alt="screen shot 2018-10-09 at 5 08 02 pm" src="https://user-images.githubusercontent.com/236943/46707336-5a279a80-cbee-11e8-8308-b66fdd5a1380.png">

If we were to break things apart by page, similar to how force does it, maybe we could create a template of sorts, with specific buckets that would grow over time:

<img width="377" alt="screen shot 2018-10-09 at 5 16 48 pm" src="https://user-images.githubusercontent.com/236943/46707352-6d3a6a80-cbee-11e8-9523-216a71c5c1a0.png">

Which is an improvement over the current `Values.ts` file since now people have some place to go if they need to add tracking to a page, with each item grouped properly. But even this requires some mental assembly / disassembly / duplication / variable assignment vs just copying / pasting a segment snippet into track in the location that’s it’s needed in our react code and being done with it:

```js
import { ActionType, Labels } from 'tracking'
import { Artwork } from 'tracking/Artwork' 

@track({
  flow: Artwork.Flow.AboutTheWork, 
  context_module: Artwork.Context.AboutTheWorkPartner,
  type: ActionType.Click,
  label: Labels.ReadMore
})
```
vs 
```js
@track({
  flow: 'Artwork about the work',
  context_module: 'About the Work (Partner)',
  type: 'Button',
  label: 'Read more',
})
```
At scale I think these separate buckets will be hard to mentally assemble at a glance vs each individual tracking unit holding all of its values in a self-contained sort of way.

Another thing we lose is the test like quality of our existing analytics flow since all of these individual properties now sit in siloed enum buckets independent of one another. If one reads down this file:

https://github.com/artsy/force/blob/master/src/desktop/analytics/artwork.js

Each tracking entry is like a “describe” / “it” testcase. It reads well, with all of the associated data sitting right side by side. If a dev needs to make an edit due to an error, its pretty fast to fix / find.

Which also raises the question, how often are we dealing with errors in our tracking, and what is the level of effort / time to fix / find? If its a problem, is the new system actually better? Would it be more or less effort / time to fix an issue having to navigate through the new complexity overhead vs the old system? Without tooling already in place to auto-generate tracking code according to this stricter form, i’m wondering what the benefit of this new system is, and if its solving a real problem that we’ve faced in our code and felt the need to revise.

A good mental complexity metric is: how well does the new system map to a “tracking sheet” value an excel spreadsheet or JIRA ticket? Like, if given a table cell with a tracking code, how long would it take to translate that default segment form into our new system, and how quickly would one be able to translate it back to the spreadsheet/ticket? Because if encountering issues we’d need to be able to quickly cross-reference things.

(Also worth referencing: https://segment.com/docs/guides/sources/can-i-see-an-example-of-a-tracking-plan)

**Comment by cavvia** [comment]
We have a schema for analytics events here, but it is not fully upto date (as you can imagine, tracking code changes manually is always a losing game):

https://docs.google.com/spreadsheets/d/17oqFli15x1SmNBu_-fPZhFifMw2BNSPJe9VdOgs1aKs/edit#gid=0

I agree we could benefit from having a 'tracking plan' approach to documentation instead, rather than trying to exhaustively document events. I will discuss this with @ani, we should be able to create one.

Going a step further, we could benefit from a schema validation step too, where analytics specifies the schema and it is enforced at some level. Our tracking provider Segment has just released a product which allows us to specify a schema and then allows for validation against that schema inline in VS Code via a javascript library:

https://segment.com/blog/introducing-protocols

That generates inline doc on the event and highlights when you're using an invalid event name or property. It would require us to do a full audit of our schema and load that into Segment however, so there's considerable upfront cost. It may pay off in the long run though, in terms of clarity for developers and schema validation throughout the code. It would also require refactoring Force for instances where it deviates from the schema we lay down, so it provides less flexibility than we have today.

I'd say we have some issues with schema consistency and with event naming consistency, but we do have a sense of what a clean schema would look like and we've explored that [here](https://docs.google.com/document/d/18IrXhzvD3M9fzgFPgyMjLSgEBI_950ytPmup7hqBrj4/edit?usp=sharing).

**Comment by alloy** [comment]
> Going a step further, we could benefit from a schema validation step too, where analytics specifies the schema and it is enforced at some level. Our tracking provider Segment has just released a product which allows us to specify a schema and then allows for validation against that schema inline in VS Code via a javascript library:
>
> https://segment.com/blog/introducing-protocols
>
> That generates inline doc on the event and highlights when you're using an invalid event name or property. It would require us to do a full audit of our schema and load that into Segment however, so there's considerable upfront cost.

@cavvia btw, I’d prefer to not have vendor specific code throughout the codebase, but rather use an abstraction (such as react-tracking).

By leveraging TypeScript we get pretty much the same type of feedback and doesn’t require spending time to load a schema into Segment, but instead in our vendor-agnostic TS schema.

**Comment by cavvia** [comment]
@alloy I completely agree with you on vendor libraries!

**Comment by damassi** [comment]
> This model doesn’t work across platforms, as it requires e.g. a queryable DOM as data container

@alloy - to be clear, I'm not suggesting that we follow force's way and I hope I didn't suggest that by my issue. I'm all for `react-tracking`. What I'm saying is that the new pattern being set down that takes an easy-peasy JS object handed over by @cavvia and breaks it apart into "buckets" spread across multiple files -- meaning, encoded enums categorized by type -- ambiguates things, introduces code complexity and doesn't necessarily solve an observed problem.

@mzikherman, @sweir27 - I tagged y'all in this issue because I know y'all have written a lot of tracking code in Force. Have you ever run into any issues? If so what were they?

Let me rephrase: The new pattern that is being laid down introduces complexity. 

This:

```js
analytics.track('Click', {
  flow: 'Artwork about the work',
  context_module: 'About the Work (Partner)',
  type: 'Button',
  label: 'Read more',
})
```

Forces whatever random dev who's been assigned the task to break things apart and track down the proper buck for these values and then drop them in: 

```js
enum Flow {
  ArtworkAboutTheWork = 'Artwork about the work',
  ...
}

// some other file / location 

enum ContextModule {
  AboutTheWorkPartner = ''About the Work (Partner)'
  ...
}

// some other file / location 

enum ActionTypes {
   Button = 'Button',
   ... 
}

// some other file / location

enum Label {
  ReadMore = 'Read more',
  ...
}
```

Which then needs to be re-assembled in the code base: 

```js
import { Flow, ContextModule, Type, Label, Actions } from 'Artsy/Analytics'

@tracking({
  flow: Flow.ArtwokrAboutTheWork,
  context_module: ContextModule.AboutTheWorkPartner,
  type: Types.Button,
  label: Label.ReadMore,
  action: Action.Click
})
export class MyComponent extends React.Component {} 
```

For a smallish app these bucks seem orderly and the example code above straight forward / simple enough, but looking at force shows us that **at scale** this is going to get really messy an confusing, requiring devs to _unsow_ and encode idiomatic segment tracking code into type-safe enums and then later _resow_ things back into their UI imports, variables, and so on. 

Along the way duplication happens since string values have to be encoded into code-usable enum `CapsCase` namings. This creates drift, duplication and complexity and its not necessary since most of these values are unique anyway. For those that aren't, we can clean up and consolidate behind something reusable (e.g., `Action.Click`).

This directory and its files contains _a lot_ of code: https://github.com/artsy/force/blob/master/src/desktop/analytics

If the issue is sharing across platforms, something like the below is cross-platform, would bring us closer to the form that is handed off to us from @cavvia and @anipetrov, be close to idiomatic segment patterns, remove complexity, and also be cross-referencable should errors arise:

```js
// analytics/artist.js 

export const artistTracking = {
  'someID': (props, state) => ({
     flow: 'Artwork about the artist',
     context_module: 'Artwork tabs',
     type: 'Tab',
     label: 'exhibition_highlights|articles|biography|about_the_work'
  }),
  'anotherID': (props, state) => ({
     flow: '...',
     context_module: '...',
     type: '...'
     label: '...
  }),
   'andAnotherID': (props, state) => {
     flow: '...',
     context_module: '...',
     type: '...'
     label: '...
  }
}
```

We then import this object into the areas that we need to track, and pass by id:

```js
import { artistTracking } from 'analytics/artist'

@track(artistTracking.someID) 
export class Foo extends React.Component {} 
```

Each of those ID's can exist in a spreadsheet, with the associated codes. The ID can be logged to the console during the validation phase so we have a paper trail. The tracking file reads linearly and as one would expect, without the need to ambiguate and disambiguate for the sake of a compiler and code-like organization (where what's needed is just a matrix of tracking codes).

Thoughts?

**Comment by alloy** [comment]
> forces whatever random dev who's been assigned the task to break things apart and track down the proper bucket for these values and then drop them in

The idea has always been that the analytics team would use the standardised names when communicating, so this doesn’t need to be as hard as that sounds.

I have never had a `analytics.track(…)` example being communicated to me before. I get sheets and have to line that up with the existing tracks.

> If the issue is sharing across platforms, something like the below is cross-platform

I don’t think the code that uses component data should be in other files, because that increases the risk of breaking data dependencies.

I don’t care for these things being an enum specifically, so another option would be to change things like action name, context etc from an enum to union strings:

```ts
type Context =
  | "MinimalCtaBanner"
  | "etc"

type ActionType =
  | "AuthImpression"
  | "etc"

type ActionName =
  | "ViewEditorial"
  | "etc"
```

That way the event can look like:

```js
@track({ 
  context_module: "MinimalCtaBanner"
})
export class MinimalCtaBanner extends React.Component {
  @track({
    action_type: "AuthImpression",
    intent: "ViewEditorial",
    trigger: "click",
  })
  componentDidMount() {
    // no op
  }
}
```

_but_ new keys that we want to be standardised in the schema definition will have to be added to the unions. However, there shouldn’t be new keys all the time, so that wouldn’t be a big issue imo.

OTOH, the same applies to the current schema definition, as long as we agree with the analytics team on how to pass these events over, so that looks to be mostly where the issue lies to me.

**Comment by damassi** [comment]
> I don’t think the code that uses component data should be in other files, because that increases the risk of breaking data dependencies.

I would like to hear others opinions, but I don't think this is a real risk in regards to analytics. We're importing things by page or context -- import `analytics/artist.js`. This works cross platform, and is roughly just data.

**Comment by alloy** [comment]
I’m not sure what you mean by “roughly just data”. What I’m referring to is that for these analytics functions to work the component needs to fetch extra data needed many cases. With the analytics functions living elsewhere it will be really easy to remove a field from the component’s query that the analytics function relies on.

**Comment by damassi** [comment]
Roughly just data = In most cases, based on looking through Force and existing precedent / past experience, what would be inside of the tracking file is just data. In cases where we need to make requests, then I think we can augment our process slightly and deal with that bridge when we get there.

**Comment by alloy** [comment]
I think you’re misunderstanding something. In tons of cases do we need to fetch data purely for analytics purposes, eg an artist slug, so we are already at that bridge.

**Comment by damassi** [comment]
Ahh I see what you mean, I thought you meant right in the middle of a tracking function call. 

> With the analytics functions living elsewhere it will be really easy to remove a field from the component’s query that the analytics function relies on.

Have we run into this issue much before? I understand concerns about safety, but there are also tradeoffs. Having to encode values in this way creates a lot of additional work, and if we haven't historically had a problem with suddenly disappearing tracking fields then we're over-engineering. Also, it would be easy enough to define an interface for areas that require additional safety.

**Comment by alloy** [comment]
> Have we run into this issue much before? I understand concerns about safety, but there are also tradeoffs. Having to encode values in this way creates a lot of additional work, and if we haven't historically had a problem with suddenly disappearing tracking fields then we're over-engineering. Also, it would be easy enough to define an interface for areas that require additional safety.

Yep, we have.

But moreover, I’m not sure I understand what in this case we’re debating. If the track just stays in the component file, like is done now, then we get all of that for free.

**Comment by damassi** [comment]
In terms of problems, I would like to hear from people who have written a lot of tracking code in Force.

And yes, I would be all for keeping tracking code co-located, e.g., 

```
@track({
  flow: 'Artwork about the work',
  context_module: 'About the Work (Partner)',
  type: 'Button',
  label: 'Read more',
})
```

**Comment by damassi** [comment]
That makes sense re REST / old setup 👍 With `track` accepting generics we'll be able to ensure that things don't break in this way.

**Comment by eessex** [comment]
Coming to this conversation a bit late, but in my experience the advantage of the new typing system is that it could prevent drift -- we had many places in publishing with hard-coded tracking in the component file, where minor changes in syntax/spaces/capitalisation/punctuation created differences between components that logically could have used the same value. 

The process of implementing analytics in the past has also often been confusing from a communications standpoint-- often they are implemented based on a spreadsheet, but those values haven't always been consistent with existing tracking language/implementations, and have resulted in passing duplicates of the same information, or introductions of new fields where they weren't actually necessary, and a long back-and-forth with usually many prs before the call is finalised.

The advantage of the new system is that we can pare down the varieties of info that could be passed to a pre-ordained list, and be able to see and compare the complete list of values in a centralised place. Maybe it makes sense to break these out a bit father rather than keeping in one schema file. I'm not picky about using enum or union strings, but found it relatively easy to put together the below example a few months ago. 

``` 
{
      action_type: ActionType.Click,
      context_module: Context.ReadMore,
      destination_path,
      subject: Subject.ReadMore,
      referrer
}
```

**Comment by damassi** [comment]
Thanks for all of the input and elaboration everyone. After some thought and implementing a number of tracking codes / updating the new schema pattern I'm going to ahead and close as a lot of my concerns have been addressed. If anyone has anything else to add please feel free to continue the thread.

**Comment by cavvia** [comment]
With regards to workflow, I supplied some tracking code but I will stick to using tracking sheets only from now on as that reinforces the tracking sheet as the single source of truth. Am happy to add identifiers to the tracking sheet if that helps to locate the event in the sheet when coming from the tracking code.

**Comment by alloy** [comment]
I’ve scheduled a meeting with @eessex @xtina-starr @cavvia @anipetrov to discuss next steps to getting to a place where the schema reflects the analytics team’s ideals and how to do communication.

**Comment by damassi** [comment]
👍 When y'all meet please try to keep things simple for future devs of varying expertise who aren't intimately familiar with Segment, TypeScript, etc. 

(i.e., a "database table" (schema) with "columns" of a particular type (segment fields) is pretty safe and understandable at a glance, and also lends itself to future iteration.) 

We can always expand on these ideas and lock things down over time. Our current tracking setup (without docs to guide the user) feels a bit over-engineered at this early stage.
